### PR TITLE
Add initial version of dict-like API (similar to `neptune-client` 1.x)

### PR DESCRIPTION
**Pull request by kgodlewski**
This is not yet ready for production use, but does not interfere with the current usage patterns.

### CODE REVIEW

File: src/neptune_scale/api/attribute.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+if TYPE_CHECKING:
+    from neptune_scale.api.run import Run
+
+
+def warn_unsupported_params(fn: Callable) -> Callable:
+    # Perform some simple heuristics to detect if a method is called with parameters
+    # that are not supported by Scale
+    warn = functools.partial(warnings.warn, stacklevel=3)
+
+    def wrapper(*args, **kwargs):  # type: ignore
```
Comment (michalsosn) [inline src/neptune_scale/api/attribute.py:29]:
```suggestion
@functools.wraps(fn)
def wrapper(*args, **kwargs): # type: ignore
```

File: src/neptune_scale/api/attribute.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+
+class AttributeStore:
+    def __init__(self, run: "Run") -> None:
+        self._run = run
+        self._attributes: Dict[str, Attribute] = {}
+
+    def __getitem__(self, key: str) -> "Attribute":
+        path = cleanup_path(key)
+        attr = self._attributes.get(key)
```
Comment (michalsosn) [inline src/neptune_scale/api/attribute.py:56]:
```
        attr = self._attributes.get(path)
```

File: src/neptune_scale/api/attribute.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        if attr is None:
+            attr = Attribute(self, path)
+            self._attributes[path] = attr
+
+        return attr
+
+    def __setitem__(self, key: str, value: ValueType) -> None:
+        # TODO: validate type if attr is already known
+        attr = self[key]
```
Comment (michalsosn) [inline src/neptune_scale/api/attribute.py:65]:
```suggestion
        path = cleanup_path(key)
        attr = self[path]
```
?

File: src/neptune_scale/api/attribute.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        wait: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        data = accumulate_dict_values(value, self._path)
+        self._store.log(metrics=data, step=step, timestamp=timestamp)
+
+    @warn_unsupported_params
+    # TODO: this should be Iterable in Run as well
+    # def add(self, values: Union[str, Iterable[str]], *, wait: bool = False) -> None:
+    def add(self, values: Union[str, Union[List[str], Set[str], Tuple[str]]], *, wait: bool = False) -> None:
```
Comment (michalsosn) [inline src/neptune_scale/api/attribute.py:132]: is it really clear that this method adds tags if you shorten add_tags to just add()?

can't you type it as `values: Union[str, Iterable[str]]`right now and copy it to tuple `self._store.log(tags_add={self._path: tuple(values)})` at the point where you're calling run.log?

File: src/neptune_scale/api/attribute.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    Optional,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
+
+if TYPE_CHECKING:
+    from neptune_scale.api.run import Run
+
```
Comment (michalsosn) [inline src/neptune_scale/api/attribute.py:22]: maybe define _all__ = []? 
There are some methods like accumulate_dict_values that seem like something that shouldn't really be public

File: src/neptune_scale/api/attribute.py (hunk 1 above)
Comment (kgodlewski) [reply src/neptune_scale/api/attribute.py:29]: I think I already added this in another branch 👍

File: src/neptune_scale/api/attribute.py (hunk 2 above)
Comment (kgodlewski) [reply src/neptune_scale/api/attribute.py:56]: This is already refactored in another branch as well :-)

File: src/neptune_scale/api/attribute.py (hunk 4 above)
Comment (kgodlewski) [reply src/neptune_scale/api/attribute.py:132]: This method name is taken from the old client, for which the API needs to be compatible. I'll add `add_/remove_tags` in another PR, deprecating this one

**Review by michalsosn — APPROVED** [review]

**Pull request merged**